lib/integrations/awsoidc/deployservice.go

```diff
@@ -0,0 +1,671 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awsoidc
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/service/ecs"
+	ecsTypes "github.com/aws/aws-sdk-go-v2/service/ecs/types"
+	"github.com/gravitational/trace"
+	"golang.org/x/exp/slices"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/retryutils"
+)
+
+var (
+	// launchTypeFargateString is the FARGATE LaunchType converted into a string.
+	launchTypeFargateString = string(ecsTypes.LaunchTypeFargate)
+	// requiredCapacityProviders contains the FARGATE type which is required to deploy a Teleport Service.
+	requiredCapacityProviders = []string{launchTypeFargateString}
+
+	// Ensure Cpu and Memory use one of the allowed combinations:
+	// https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html
+	taskCPU = "512"
+	taskMem = "1024"
+
+	// taskAgentContainerName is the name of the container to run within the Task.
+	// Each task supports multiple containers, but, currently, there's only one being used.
+	taskAgentContainerName = "teleport-service"
+
+	// oneAgent is used to define the desired agent count when creating a service.
+	oneAgent = int32(1)
+
+	// serviceForceDeletion indicates that the service must be deleted even if it has running tasks.
+	serviceForceDeletion = true
+)
+
+const (
+	// teleportContainerImageFmt is the Teleport Container Image to be used
+	teleportContainerImageFmt = "public.ecr.aws/gravitational/teleport-distroless:%s"
+
+	// clusterStatusActive is the string representing an ACTIVE ECS Cluster.
+	clusterStatusActive = "ACTIVE"
+	// clusterStatusInactive is the string representing an INACTIVE ECS Cluster.
+	clusterStatusInactive = "INACTIVE"
+	// clusterStatusProvisioning is the string representing an PROVISIONING ECS Cluster.
+	clusterStatusProvisioning = "PROVISIONING"
+	// clusterStatusProvisioningWaitTime defines for how long should the client wait for the Cluster to become available.
+	clusterStatusProvisioningWaitTime = 30 * time.Second
```

Comment:
Isn't that too short?

Reply:
From my tests we never see the Provisioning status.
We are just being extra careful here
More discussion here https://github.com/gravitational/teleport/pull/27035#discussion_r1211917793